server/src/main/java/org/elasticsearch/health/metadata/HealthMetadataService.java

```diff
@@ -53,25 +54,29 @@ public class HealthMetadataService {
     private volatile boolean enabled;
 
     // Signifies that a node has been elected as master, but it was not able yet to publish its health metadata for
-    // other reasons for example not all nodes of the cluster are 8.4.0 or newer
-    private volatile boolean readyToPublish = false;
+    // other reasons for example not all nodes of the cluster are 8.5.0 or newer
+    private volatile boolean readyToPublishAsMaster = false;
     // Allows us to know if this node is the elected master without checking the cluster state, effectively protecting
     // us from checking the cluster state before the cluster state is initialized
     private volatile boolean isMaster = false;
+    private volatile HealthMetadata previousHealthMetadata;
+    private volatile HealthMetadata healthMetadata;
```

Comment:
I find this name a bit misleading. When I first read this I thought this variable will contain the health metadata currently shared in the cluster state, and the previous the last value before this. However, after some more reading I think that these variable represent the health metadata that would be stored in the cluster if this node was the master. Right?

That got me thinking, with rollover we retrieve the current values from the cluster settings every time. If I am not mistaken our problem here is when a master is freshly elected or the cluster is ublocked how to get the correct settings. Could we maybe use something like this: `clusterSettings.get(BLAH_SETTING)` at that moment? This was we would have so much less booking to do. Right? I am not 100% sure this would work, but it feels like it should. Maybe the settings in isolation are not a good option, but the combo with cluster settings maybe is.

Reply:
Discussed offline that reading directly from the cluster settings is not the way to do this :). Thanks for the explanation @HiDAl .

So, let's rethink this:

- Maybe we can name the `healthMetadata` differently, like `localHealthSettings` or `localHealthMetadata` and explain in an one line comment why they are there.
- I like that we always listen to the cluster setting updates and update these variable no matter, if there is a block, or if this is enabled or not. Otherwise, we will miss updates.
-  I am wondering if these listeners could choose to trigger a request to update the `healthMetadata` in the cluster state if all the conditions hold, what do you think?
- When we become a master and we are ready to publish we publish the new health metadata.
- Maybe we can remove the check if health metadata has changed altogether, the cluster state task already does that, if I am not mistaken, right? Then we only trigger an update if we just became the master. What do you think?

I am trying to reduce the state in memory and the paths that lead to update.

Reply:
> Maybe we can name the healthMetadata differently, like localHealthSettings or localHealthMetadata and explain in an one line comment why they are there.

++ 

> I like that we always listen to the cluster setting updates and update these variable no matter, if there is a block, or if this is enabled or not. Otherwise, we will miss updates.
++
> I am wondering if these listeners could choose to trigger a request to update the healthMetadata in the cluster state if all the conditions hold, what do you think?

Not sure if I correctly understood this: From my perspective, they are already doing this, but transitively: trigger the change, update the in-memory stuff, a cluster state is triggered, we post the up-to-date health metadata. right?

> When we become a master and we are ready to publish we publish the new health metadata.

yes

> Maybe we can remove the check if health metadata has changed altogether, the cluster state task already does that, if I am not mistaken, right? Then we only trigger an update if we just became the master. What do you think?

I though about this, but wasn't sure. Let me think a bit more about it.

Reply:
yes, you're right, it's safe to have only 1 object in memory and let the `UpsertHealthMetadataTask.execute` do the magic of checking whether it should update or not the `ClusterState`.

Reply:
pushed :) thanks for your comment. Now it's easier to understand